src/python/pants/backend/nfpm/native_libs/deb/rules.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2025 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import importlib.metadata
+import json
+import logging
+import sys
+from collections.abc import Iterable, Mapping
+from dataclasses import dataclass, replace
+from pathlib import PurePath
+
+from pants.backend.python.util_rules.pex import PexRequest, VenvPexProcess, create_venv_pex
+from pants.backend.python.util_rules.pex_environment import PythonExecutable
+from pants.backend.python.util_rules.pex_requirements import PexRequirements
+from pants.engine.fs import CreateDigest, FileContent
+from pants.engine.internals.native_engine import UnionRule
+from pants.engine.internals.selectors import concurrently
+from pants.engine.intrinsics import create_digest, execute_process
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import Rule, collect_rules, implicitly, rule
+from pants.init.import_util import find_matching_distributions
+from pants.util.logging import LogLevel
+from pants.util.resources import read_resource
+from pants.version import VERSION
+
+logger = logging.getLogger(__name__)
+
+_NATIVE_LIBS_DEB_PACKAGE = "pants.backend.nfpm.native_libs.deb"
+_SEARCH_FOR_SONAMES_SCRIPT = "search_for_sonames.py"
+_PEX_NAME = "native_libs_deb.pex"
+
+
+@dataclass(frozen=True)
+class DebSearchForSonamesRequest:
+    distro: str
+    distro_codename: str
+    debian_arch: str
+    sonames: tuple[str, ...]
+    from_best_so_files: bool
+
+    def __init__(
+        self,
+        distro: str,
+        distro_codename: str,
+        debian_arch: str,
+        sonames: Iterable[str],
+        *,
+        from_best_so_files: bool = False,
+    ):
+        object.__setattr__(self, "distro", distro)
+        object.__setattr__(self, "distro_codename", distro_codename)
+        object.__setattr__(self, "debian_arch", debian_arch)
+        object.__setattr__(self, "sonames", tuple(sorted(sonames)))
+        object.__setattr__(self, "from_best_so_files", from_best_so_files)
+
+
+@dataclass(frozen=True)
+class DebPackagesPerSoFile:
+    so_file: str
+    packages: tuple[str, ...]
+
+    def __init__(self, so_file: str, packages: Iterable[str]):
+        object.__setattr__(self, "so_file", so_file)
+        object.__setattr__(self, "packages", tuple(sorted(packages)))
+
+
+_TYPICAL_LD_PATH_PATTERNS = (
+    # platform specific system libs (like libc) get selected first
+    # "/usr/local/lib/*-linux-*/",
+    "/lib/*-linux-*/",
+    "/usr/lib/*-linux-*/",
+    # Then look for a generic system libs
+    # "/usr/local/lib/",
+    "/lib/",
+    "/usr/lib/",
+    # Anything else has to be added manually to dependencies.
+    # These rules cannot use symbols or shlibs metadata to inform package selection.
+)
```

Comment:
This is a simplified heuristic, compared to the actual logic used by `/lib/ld-linux.so` and similar dynamic ELF loaders.

I commented out the `/usr/local` variants because packages available via the debian/ubuntu package search APIs are unlikely to install anything in this sysadmin-managed directory.